src/main/java/edu/indiana/dlib/amppd/service/impl/MediaServiceImpl.java

```diff
@@ -302,6 +309,78 @@ else if (asset instanceof CollectionSupplement) {
 		}
 		return asset;
 	}
-
+	
+	public ItemSearchResponse findItemOrFile(String keyword, String mediaType) {
+		log.info("Executing file search in media service");
+		ItemSearchResponse response = new ItemSearchResponse();
+		ArrayList<ItemSearchResult> rows = new ArrayList<ItemSearchResult>();
+		try {
+			List<Primaryfile> matchedFiles = primaryfileRepository.findByItemOrFileName(keyword);
+			ItemSearchResult result = new ItemSearchResult();;
+			Map <String, String>primaryFilerows = new HashMap<String, String>();
+			log.debug("the first object is:"+matchedFiles.get(0).toString());
+			long curr_item_id = 0;
+			for(Primaryfile p : matchedFiles) {
+				//reset if the current item is a new entry
+				if(p.getItem().getId() != curr_item_id && primaryFilerows.size()>0) {
+					log.info("Now new item id:"+p.getItem().getId()+" curr item id:"+curr_item_id);
+					result.setPrimaryFiles(primaryFilerows);
+					rows.add(result);
+					result = new ItemSearchResult();
+					primaryFilerows = new HashMap<String, String>();
+					
+				}
+				String mime_type = getMediaTypeFromJson(p);
+				if(mime_type!=null && !mediaType.contentEquals("000"))
+				{
+					if((mime_type.contains("audio") && mediaType.substring(0, 1).contentEquals("1")) 
+							|| (mime_type.contains("video") && mediaType.substring(1, 2).contentEquals("1")) 
+							|| (!mime_type.contains("video") && !mime_type.contains("audio") && mediaType.contentEquals("001"))){
+						curr_item_id = p.getItem().getId();
+						result.setItemName(p.getItem().getName());
+						primaryFilerows.putIfAbsent(p.getId().toString(), p.getOriginalFilename());
+					}
+				}
+				else {
+					curr_item_id = p.getItem().getId();
+					result.setItemName(p.getItem().getName());
+					primaryFilerows.putIfAbsent(p.getId().toString(), p.getOriginalFilename());
+
+				}
+			}
+			//add the last item to the rows
+			if(primaryFilerows.size()>0) {
+				result.setPrimaryFiles(primaryFilerows);
+				rows.add(result);
+				response.setRows(rows);
+				response.setSuccess(true);
+			}
+			response.setSuccess(false);
+			response.setError("No primary file found");
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+			response.setError(e.getMessage());
+			response.setSuccess(false);
+		}
+		return response;
+	}
+	
+	public String getMediaTypeFromJson(Primaryfile p) {
+		String mime_type = new String();
+		String media_type = p.getMediaInfo();
+		try {
+			if(media_type!=null) {
+				JSONObject jsonObject = new JSONObject(media_type);
+				JSONObject jsonObject_container = new JSONObject(jsonObject.getString("container"));
+				mime_type = jsonObject_container.getString("mime_type");
+				log.info("====>>>>>MIME TYPE IS:"+jsonObject_container.getString("mime_type"));
+			}
+		} catch (JSONException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		return mime_type;
+	}
 
 }
```

Comment:
We can leave this as is for now. Ideally we'd want to impl the search at the data level using custom query in repository, then we don't need extra service or controller., mediaType should be queryable with postgresSql's extension on JSONB. this is our original consideration to use JSONB - to improve performance. And if the query is too complicated or the field is frequently used, we can extract it into a separate field in DB table. The point is to avoid parsing the JSON each time in code.